v4/service.go

```diff
@@ -0,0 +1,81 @@
+package suture
+
+import (
+	"context"
+	"errors"
+)
+
+/*
+Service is the interface that describes a service to a Supervisor.
+
+Serve Method
+
+The Serve method is called by a Supervisor to start the service.
+The service should execute within the goroutine that this is
+called in, that is, it should not spawn a "worker" goroutine.
+If this function either returns error or panics, the Supervisor
+will call it again, unless it returns special error sentinel values
+
+A Serve method SHOULD do as much cleanup of the state as possible,
+to prevent any corruption in the previous state from crashing the
+service again.
+
+The Supervisor will pass a ctx down to the service. That context
+can not be independently used to cancel the service, as the supervisor
+will not notice the context is cancelled. Use the Remove method on the
+Supervisor to remove it properly, or have the service return
+ErrDoNotRestart.
+
+The error returned by the service, if any, will be part of the log
+message generated for it. There are two distinguished errors a
+Service can return: ErrDoNotRestart indicates that the service should
+not be restarted and removed from the supervisor entirely.
+
+ErrTerminateTree indicates that the entire supervision tree should be
+restarted.
```

Comment:
Restarted or terminated?
The comment says restarted, the error says terminated.

Atleast in syncthings case, we're loking for termination, as we want to unwind everything in case of a fatal database error or something along those lines.